### PR TITLE
Python: Fix MCP client infinite loop when server does not support optional ping method

### LINKED ISSUE #4940: Python: [Bug]: MCP client connection goes into infinite loop if MCP server doesn't support "ping" method
### Description

I have used the MCPStreamableHTTPTool to build a client that connects to snowflake managed MCP server. The connection requests were going into infinite loops. Upon reviewing the code implementation of MCPTool, I realized that the code was written under the assumption that every MCP server implements the "ping" method. However, the snowflake managed MCP server doesn't implement this method. 

When I make a call to **load_tools**, it first checks connection using _ensure_connected that in turn uses **ping**. When **ping** fails, it tries to **connect** again.

if we want to make use of **call_tool**, we must needs keep the **load_tools=True** while initializing. If **load_tools=True**, in **connect** method, it will try to call **load_tools** which in turn calls **_ensure_connected** that uses **ping** and declares that the connection is invalid and tries to reconnect using the **connect** method.

The following leads to an infinite loop
connect=>load_tools or load_prompts => _ensure_connected => ping fails => connect

### Code Sample

```markdown
import asyncio
import os
from agent_framework import MCPStreamableHTTPTool
import httpx

headers = {
    "Authorization": f"Bearer {os.getenv('SNOWFLAKE_MCP_TOKEN')}",
    "Content-Type": "application/json",
}
url = os.getenv("SNOWFLAKE_MCP_URL")

async_client = httpx.AsyncClient(headers=headers, base_url=url)
client = MCPStreamableHTTPTool(
    name="snowflake_tool",
    description="Tool to execute SQL queries against Snowflake and stream results back.",
    url=url,
    http_client=async_client,
)

asyncio.run(client.load_tools())
```

### Error Messages / Stack Traces

```markdown
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:agent_framework._mcp:MCP connection invalid or closed. Reconnecting...
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 200 OK"
INFO:mcp.client.streamable_http:Negotiated protocol version: 2025-11-25
INFO:httpx:HTTP Request: POST https://*************************************************************************** "HTTP/1.1 202 Accepted"
```

### Package Versions

agent-framework-core==1.0.0rc3

### Python Version

Python 3.12.10

### Additional Context

I feel _ensure_connected can call the **ping** method. If the response is an **MCPError** that has "Unsupported method: ping" in its error body, it can be treated as a live connection or some other logic should be implemented here.

### PR DESCRIPTION
**Pull request by giles17**
### Motivation and Context

MCP servers are not required to implement the `ping` method (it is optional per the MCP spec). When connecting to such a server, `_ensure_connected` treated ping failure as a dead connection, triggering reconnection which re-invoked `load_tools`/`load_prompts`, creating an infinite `connect → load_tools → _ensure_connected → ping fails → connect` loop.

Fixes #4940

### Description

The root cause was that `_ensure_connected` did not distinguish between a genuinely dead connection and a server that simply doesn't implement the optional ping method. The fix catches `McpError` with the JSON-RPC "Method not found" error code (`-32601`) from `send_ping` and treats it as a valid, live connection—skipping reconnection. Any other `McpError` code still triggers the existing reconnection logic. Tests verify that ping-unsupported servers no longer cause infinite loops, that `-32601` is treated as connected, and that other MCP errors still reconnect normally.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by giles17's agent

### CODE REVIEW

File: python/packages/core/agent_framework/_mcp.py [hunk 1]
```diff
@@ -1039,8 +1039,28 @@ async def _ensure_connected(self) -> None:
...
+        METHOD_NOT_FOUND = -32601
+
         try:
             await self.session.send_ping()  # type: ignore[union-attr]
+        except McpError as mcp_exc:
+            # If the server responds with "Method not found", it means the
+            # connection is alive but the server doesn't implement the optional
+            # ping method (per the MCP spec, ping is optional). Treat as connected.
+            if mcp_exc.error.code == METHOD_NOT_FOUND:
+                logger.debug("MCP server does not support ping (method not found). Connection is valid.")
```
Comment (giles17) [inline python/packages/core/agent_framework/_mcp.py:1064]: This avoids the reconnect loop, but leaves `_ensure_connected()` on an exception-as-control-flow path for every server that legitimately omits the optional `ping` RPC. Both `load_prompts()` (line 904) and `load_tools()` (line 954) invoke `_ensure_connected()` inside pagination loops, so ping-less servers pay a `McpError(-32601)` round-trip on every refresh. Consider remembering once that `ping` is unsupported (e.g., a `self._ping_supported` flag) so these servers stay on a normal path.

File: python/packages/core/tests/core/test_mcp.py [hunk 2]
```diff
@@ -3662,6 +3662,61 @@ async def test_ensure_connected_wraps_reconnect_failure() -> None:
...
+    """McpError with a code other than -32601 should still trigger reconnection."""
+    tool = MCPTool(name="test_tool")
+    tool.session = Mock(
+        send_ping=AsyncMock(side_effect=McpError(types.ErrorData(code=-32603, message="Internal error")))
+    )
+
+    with patch.object(tool, "connect", AsyncMock()) as mock_connect:
+        await tool._ensure_connected()
+
+    mock_connect.assert_awaited_once_with(reset=True)
```
Comment (giles17) [inline python/packages/core/tests/core/test_mcp.py]: This test covers successful reconnection after a non-method-not-found McpError, but there's no corresponding test for the failure case (when `connect(reset=True)` raises). The `except McpError` branch has its own `try/except` for reconnection (lines 1057-1063) that wraps failures in `ToolExecutionException`. Consider adding a test analogous to `test_ensure_connected_wraps_reconnect_failure` but using `McpError(code=-32603)` as the ping side-effect.

```suggestion
    mock_connect.assert_awaited_once_with(reset=True)


async def test_ensure_connected_wraps_reconnect_failure_on_mcp_error() -> None:
    ""McpError with non-method-not-found code + reconnect failure should raise ToolExecutionException."
    tool = MCPTool(name="test_tool")
    tool.session = Mock(
        send_ping=AsyncMock(side_effect=McpError(types.ErrorData(code=-32603, message="Internal error")))
    )

    with (
        patch.object(tool, "connect", AsyncMock(side_effect=RuntimeError("still closed")),
        pytest.raises(ToolExecutionException, match="Failed to establish MCP connection"),
    ):
        await tool._ensure_connected()
```

**Review by giles17 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 4 | **Confidence:** 90%

### ✓ Correctness
> The change correctly handles the case where an MCP server does not support the optional ping method by catching McpError with code -32601 and treating it as a valid connection. The exception handling order is correct (McpError before Exception), the ErrorData.code attribute access is valid, and the tests cover the key scenarios. No correctness issues found.

### ✓ Security Reliability
> The change correctly handles the case where an MCP server does not support the optional ping method (JSON-RPC -32601 'Method not found'). The exception handling order is correct (McpError before Exception), the attribute access pattern (mcp_exc.error.code) is verified against the McpError class definition, and the logic is sound — any response (even an error) from the server proves the connection is alive. No security or reliability issues found.

### ✓ Test Coverage
> The new tests adequately cover the main scenarios for the McpError handling in `_ensure_connected`: method-not-found treated as connected, non-method-not-found triggering reconnection, and the integration test verifying no infinite loop. One edge case is missing: reconnection failure after a non-method-not-found McpError (the duplicated reconnect logic in the `except McpError` branch at lines 1057-1063 is not tested for the failure path, unlike the identical logic in the generic `except Exception` branch which is tested by `test_ensure_connected_wraps_reconnect_failure`).

### ✗ Design Approach
> The change fixes the immediate reconnect loop for servers that do not implement the optional `ping` RPC, but it still treats that unsupported optional method as the steady-state health-check path. Because `_ensure_connected()` continues to call `send_ping()` every time and only special-cases `-32601` after the exception is raised, ping-less servers will remain on an exception-driven path during every paginated `load_tools()` / `load_prompts()` refresh instead of learning once that ping is unsupported or using a required request as the liveness signal.

### Flagged Issues
- [ ] `_ensure_connected()` still unconditionally calls `send_ping()` as its liveness probe even though `ping` is optional. Because `load_prompts()` and `load_tools()` call `_ensure_connected()` inside pagination loops, a server without `ping` will raise `McpError(code=-32601)` on every page load indefinitely. The better approach is to detect once that ping is unsupported and skip future probes for that session, or use a non-optional MCP operation as the connectivity check.

---
*Automated review by giles17's agents*

**Comment by moonbox3** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py">_mcp.py</a></td><td>626</td><td>30</td><td>95%</td><td><a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L153">153</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L296">296</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L356-L357">356&ndash;357</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L486">486</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L548">548</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L561">561</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L563-L566">563&ndash;566</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L585-L586">585&ndash;586</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L599-L602">599&ndash;602</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L604-L605">604&ndash;605</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L609">609</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L673-L675">673&ndash;675</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L1034">1034</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L1169">1169</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L1221-L1222">1221&ndash;1222</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L1225">1225</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L1243">1243</a>, <a href="https://github.com/microsoft/agent-framework/blob/6d599b55fe40479130723dfd6decc0fd8a0e665f/packages/core/agent_framework/_mcp.py#L1696">1696</a></td></tr><tr><td><b>TOTAL</b></td><td><b>33642</b></td><td><b>3912</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 6569 | 30 :zzz: | 0 :x: | 0 :fire: | 1m 47s :stopwatch: |

File: python/packages/core/agent_framework/_mcp.py [hunk 3]
```diff
@@ -1039,8 +1039,28 @@ async def _ensure_connected(self) -> None:
         Raises:
             ToolExecutionException: If reconnection fails.
         """
+        from mcp.shared.exceptions import McpError
+
+        # JSON-RPC error code for "Method not found"
+        METHOD_NOT_FOUND = -32601
```
Comment (eavanvalkenburg) [inline python/packages/core/agent_framework/_mcp.py]: no point in having this constant inside the method body, either move as a constant at the top of the file, or just set this at the place it is used (that would be my preference, since it's used just once)
